### ISSUE TITLE
ui.Slider: intial value sometimes (randomly) not honored 

### ISSUE THREAD

**Issue by eudoxos** [zoofIO/flexx#212]
With the following MWE and flexx from git:

```
from flexx import app, ui
class Sliderbug(ui.Widget):
    def init(self):
        ui.Slider(min=3,max=10,value=6,step=1.)
if __name__=='__main__':
    app.launch(Sliderbug)
    app.run()

```

I sometimes get the initial value wrong, about 50%-50%, running exactly the same code. I can also verify that in the first (buggy) case the slider gives "1" as its value (rather than 3, which is the minimum):

![slider-bug](https://cloud.githubusercontent.com/assets/1029876/18514001/06a3771e-7a90-11e6-8531-67dd2413cd59.png)

![slider-ok](https://cloud.githubusercontent.com/assets/1029876/18514007/0e0e735a-7a90-11e6-8e58-eb080c3e3b7f.png)

**Comment by Korijn** [comment]
Looks like a race condition in the application of the kwargs, as dicts are unordered. Applying value is probably validated wrt min/max, so that can lead to strange behaviour.

**Comment by eudoxos** [comment]
Yes, true, that looks like the cause (python 3.6's dict is ordered by default, FWIW). If I set value after the object is constructe, the issue is no more present. Is it something worth fixing in the ctor (reordering the kwargs so that value comes last) or should I use the 2-line workaround in my code?

**Comment by Korijn** [comment]
I think you can label event connectors to enforce order, but I'm not sure about properties...

**Comment by almarklein** [comment]
Interesting case :)

The relevant code is here: https://github.com/zoofIO/flexx/blob/master/flexx/event/_hasevents.py#L156-L165, and shows that indeed the order of properties passed via the constructor is not deterministic. If we change that to process properties in alphabetic order (which is also the order of `__properties__`) then this particular issue is fixed (`"value" > "min"`). 

However, there might be cases where a certain order of property initialization is also required, and choosing the name of properties to influence that order is hardly ideal :P.

One possible solution is that we use `__properties__` as the defining order. In case a class needs a particular order, its `Foo.__properties__` list can be modified right after the class definition (and don't forget to also change `Foo.JS.__properties__`). A bit hacky, but probably good enough for now (or for ever if such cases are rare).

**Comment by Korijn** [comment]
I wonder if the HTML5 spec has anything to say (and possibly a solution) about this...

**Comment by almarklein** [comment]
Yeah, but the problem lies in the default values. If `value` is set first, it uses the default `min` and `max` to clamp itself.

**Comment by Korijn** [comment]
Took me a minute, but here's what the [HTML5 spec has to say about default behavior, and input validation](https://www.w3.org/TR/html5/forms.html#range-state-(type=range)):

> The value sanitization algorithm is as follows: If the value of the element is not a valid floating-point number, then set it to a valid floating-point number that represents the default value.
> 
> The min attribute, if specified, must have a value that is a valid floating-point number. The default minimum is 0. The max attribute, if specified, must have a value that is a valid floating-point number. The default maximum is 100.
> 
> The default value is the minimum plus half the difference between the minimum and the maximum, unless the maximum is less than the minimum, in which case the default value is the minimum.
> 
> When the element is suffering from an underflow, the user agent must set the element's value to a valid floating-point number that represents the minimum.
> 
> When the element is suffering from an overflow, if the maximum is not less than the minimum, the user agent must set the element's value to a valid floating-point number that represents the maximum.
> 
> The step scale factor is 1. The default step is 1 (allowing only integers, unless the min attribute has a non-integer value).
> 
> When the element is suffering from a step mismatch, the user agent must round the element's value to the nearest number for which the element would not suffer from a step mismatch, and which is greater than or equal to the minimum, and, if the maximum is not less than the minimum, which is less than or equal to the maximum, if there is a number that matches these constraints. If two numbers match these constraints, then user agents must use the one nearest to positive infinity.
> 
> For example, the markup `<input type="range" min=0 max=100 step=20 value=50>` results in a range control whose initial value is 60.

**Comment by almarklein** [comment]
Thanks, but I meant that Flexx has default values of the slider's min and max, which cause a problem. So either it should not have default values (which is probably annoying) or we need to force an order. (I think we should have a deterministic order regardless).

**Comment by almarklein** [comment]
This was fixed in #328 by making the order consistent. In this particular case, setting the properties in alphabetic order happens to fix things. In cases where this is not the case, we need a different approach, like what @Korijn suggested with smarter setters. I am not aware of such a case at this point though, so let's tackle that when needed.
